src/train.py

```diff
@@ -104,62 +104,169 @@ def masked_mse(gold_mel, pred_mel, mel_mask):
     mel_aug = specaugment(mel, mel_len)
-    text_pred = model.asr(text, text_len, mel_aug, mel_len).permute(0, 2, 1)
+    if use_dis_loss:
+        pre_pred, post_pred, stop_pred, t_hid = model.tts(text, text_len, mel, mel_len, ret_enc_hid=use_dis_loss)
+        t_d_loss = discriminator_hidden_to_loss(model, t_hid, 'speech')
 
-    tts_loss = speech_loss(gold_mel.to(DEVICE), gold_stop.to(DEVICE), pred, mel_len, stop_pred)
+        text_pred, s_hid = model.asr(text, text_len, mel_aug, mel_len, ret_enc_hid=use_dis_loss)
+        text_pred = text_pred.permute(0, 2, 1)
+        s_d_loss = discriminator_hidden_to_loss(model, s_hid, 'text')
+    else:
+        pre_pred, post_pred, stop_pred = model.tts(text, text_len, mel, mel_len)
+        text_pred = model.asr(text, text_len, mel_aug, mel_len).permute(0, 2, 1)
+
+    tts_loss = speech_loss(gold_mel.to(DEVICE), gold_stop.to(DEVICE), pre_pred, post_pred, mel_len, stop_pred)
     asr_loss = text_loss(gold_char.to(DEVICE), text_pred)
+    if use_dis_loss:
+        return asr_loss, tts_loss, t_d_loss, s_d_loss
     return asr_loss, tts_loss
 
-def crossmodel_step(model, batch):
+def crossmodel_step(model, batch, use_dis_loss=False):
     #NOTE: not sure if this will fail bc multiple grads on the model...
     x, y = batch
     text, mel, text_len,  mel_len  = x
     gold_char, gold_mel, gold_stop = y
 
     # Do speech!
-    pred, stop_pred = model.cm_speech_in(mel, mel_len)
-    s_cm_loss = speech_loss(gold_mel.to(DEVICE), gold_stop.to(DEVICE), pred, mel_len, stop_pred)
+    if use_dis_loss:
+        pre_pred, post_pred, stop_pred, cm_t_hid = model.cm_speech_in(mel, mel_len, ret_enc_hid=use_dis_loss)
+        cm_t_d_loss = discriminator_hidden_to_loss(model, cm_t_hid, 'speech')
+    else:
+        pre_pred, post_pred, stop_pred = model.cm_speech_in(mel, mel_len)
+    s_cm_loss = speech_loss(gold_mel.to(DEVICE), gold_stop.to(DEVICE), pre_pred, post_pred, mel_len, stop_pred)
 
     # Now do text!
-    text_pred = model.cm_text_in(text, text_len).permute(0, 2, 1)
+    if use_dis_loss:
+        text_pred, cm_s_hid = model.cm_text_in(text, text_len, ret_enc_hid=args.use_discriminator)
+        text_pred = text_pred.permute(0, 2, 1)
+        cm_s_d_loss = discriminator_hidden_to_loss(model, cm_s_hid, 'text')
+    else:
+        text_pred = model.cm_text_in(text, text_len).permute(0, 2, 1)
     t_cm_loss = text_loss(gold_char.to(DEVICE), text_pred)
+
+    if use_dis_loss:
+        return t_cm_loss, s_cm_loss, cm_t_d_loss, cm_s_d_loss
     return t_cm_loss, s_cm_loss
 
+def discriminator_hidden_to_loss(model, hid, target_type):
+    d_in = hid[-1]
+    d_out = model.discriminator(d_in)
+    target = discriminator_target(d_out, target_type)
+    d_loss = discriminator_loss(d_out, target)
+    return d_loss
+
+def discriminator_step(model, batch):
+    x, _ = process_batch(batch)
+    text, mel, text_len, mel_len = x
+
+    # text
+    t_enc_out, _ = model.text_m.encode(text, text_len)
```

Comment:
Can we call these in a context manager, depending on who is getting updated?  So if it is discriminator's turn:
`
if train_d:
  with torch.no_grad():
       t_enc_out, _ = model.text_m.encode(text, text_len)
else:
       t_enc_out, _ = model.text_m.encode(text, text_len)
`
Helps to make sure we don't give gradients to our encoder, and makes the code run faster!